src/spoke/Spoke.sol

```diff
@@ -290,64 +292,112 @@ abstract contract Spoke is ISpoke, Multicall, NoncesKeyed, AccessManagedUpgradea
+    RepayVars memory vars;
+
     Reserve storage reserve = _getReserve(reserveId);
     UserPosition storage userPosition = _userPositions[onBehalfOf][reserveId];
+    PositionStatus storage positionStatus = _positionStatus[onBehalfOf];
     _validateRepay(reserve);
 
-    IHubBase.PremiumDelta memory premiumDelta = IHubBase.PremiumDelta({
-      sharesDelta: -userPosition.premiumShares.toInt256(),
-      offsetDeltaRay: -userPosition.premiumOffsetRay.toInt256(),
-      accruedPremiumRay: 0, // populated below
-      restoredPremiumRay: 0 // populated below
-    });
+    vars.hub = reserve.hub;
+    vars.assetId = reserve.assetId;
+    vars.premiumShares = userPosition.premiumShares;
+    vars.premiumOffsetRay = userPosition.premiumOffsetRay;
+    vars.drawnIndex = vars.hub.getAssetDrawnIndex(vars.assetId);
 
-    uint256 drawnDebtRestored;
-    uint256 realizedPremiumRay;
-    (drawnDebtRestored, , realizedPremiumRay, premiumDelta.accruedPremiumRay) = _getUserDebt(
-      reserve.hub,
-      reserve.assetId,
-      userPosition
+    (vars.drawnDebtRestored, , vars.realizedPremiumRay, vars.accruedPremiumRay) = _getUserDebt(
+      userPosition,
+      vars.drawnIndex
     );
 
-    (drawnDebtRestored, premiumDelta.restoredPremiumRay) = _calculateRestoreAmount(
-      drawnDebtRestored,
-      realizedPremiumRay + premiumDelta.accruedPremiumRay,
+    (vars.drawnDebtRestored, vars.premiumRayRestored) = _calculateRestoreAmount(
+      vars.drawnDebtRestored,
+      vars.realizedPremiumRay + vars.accruedPremiumRay,
       amount
     );
 
-    uint256 premiumDebtRestored = premiumDelta.restoredPremiumRay.fromRayUp();
-    reserve.underlying.safeTransferFrom(
-      msg.sender,
-      address(reserve.hub),
-      drawnDebtRestored + premiumDebtRestored
-    );
-    uint256 restoredShares = reserve.hub.restore(reserve.assetId, drawnDebtRestored, premiumDelta);
+    vars.restoredShares = vars.hub.previewRestoreByAssets(vars.assetId, vars.drawnDebtRestored);
+    vars.newDrawnShares = userPosition.drawnShares.uncheckedSub(vars.restoredShares);
+    IHubBase.PremiumDelta memory premiumDelta = _getPremiumDelta({
+      oldPremiumShares: vars.premiumShares,
+      oldPremiumOffsetRay: vars.premiumOffsetRay,
+      newDrawnShares: vars.newDrawnShares,
+      drawnIndex: vars.drawnIndex,
+      riskPremium: positionStatus.riskPremium
+    });
+    premiumDelta.restoredPremiumRay = vars.premiumRayRestored;
 
-    userPosition.applyPremiumDelta(premiumDelta);
-    userPosition.drawnShares -= restoredShares.toUint120();
-    if (userPosition.drawnShares == 0) {
-      _positionStatus[onBehalfOf].setBorrowing(reserveId, false);
-    }
+    vars.totalDebtRestored = vars.drawnDebtRestored + premiumDelta.restoredPremiumRay.fromRayUp();
+    reserve.underlying.safeTransferFrom(msg.sender, address(vars.hub), vars.totalDebtRestored);
+    vars.hub.restore(vars.assetId, vars.drawnDebtRestored, premiumDelta);
```

Comment:
Can we use the hub returned shares amount here to verify the preview amount used before was indeed correct ?